File: 39-combination-sum/combination-sum.py
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        res = []
        def myFun(index,arr,s):
            if s == target:
                res.append(arr[:])
                return
            if s > target or index == len(candidates):
                return
            arr.append(candidates[index])
            myFun(index, arr,s+candidates[index])
            arr.pop(-1)
            myFun(index+1,arr,s)
        myFun(0,[],0)
        return res

File: 39-combination-sum/combination-sum.java
class Solution {
    public List<List<Integer>> result;
    int [] candidatesG;
    int targetG;
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        result = new ArrayList<>();
        candidatesG = candidates;
        targetG = target;
        fun(0,new ArrayList<>(),0);
        return result;
    }
    public void fun(int index, List<Integer> curr,int sum){
        if(sum == targetG){
            result.add(new ArrayList<>(curr));
            return;
        }
        if(sum > targetG || index == candidatesG.length){
            return;
        }
        curr.add(candidatesG[index]);
        fun(index,curr,sum + candidatesG[index]);
        curr.remove(curr.size()-1);
        fun(index + 1, curr,sum);
    }
}